numpy/lib/npyio.py

```diff
@@ -1108,7 +1101,8 @@ def read_data(chunk_size):
         # Read until we find a line with some values, and use it to determine
         # the need for decoding and estimate the number of columns.
         for first_line in fh:
-            ncols = len(usecols or split_line(first_line))
+            ncols = len(usecols
+                        or split_line(_decode_line(first_line, encoding)))
```

Comment:
Do you think we can eliminate the use of `decode_line` on lines in the file entirely? Either by detecting whether the file is text right at the beginning:
```
needs_decode = not isinstance(fh, io.TextIOBase)
```
or pulling off the first line much earlier:
```
for first_line in fh:
    needs_decode = isinstance(first_line, bytes)
    fh = itertools.chain([first_line], fh)
    break
else:
    needs_decode = False
```

Reply:
This should be possible, but makes things a bit more complicated because then you also need (in the later case where you need to pull lines out to know the type) to handle the possibility that the generator is empty, and then emit the empty-file warning in that case.  So from a readability PoV it may not be a big net plus, and of course speedwise it shouldn't matter at all, so let's perhaps defer that to another time?

Reply:
The only reason I raise it is that this refactor leads to the `latin1` encoding cropping up in multiple places.

> This should be possible, but makes things a bit more complicated because then you also need (in the later case where you need to pull lines out to know the type) to handle the possibility that the generator is empty, and then emit the empty-file warning in that case.

I don't think you do - you can still defer emitting the that warning until you try and count the columns of the file; you can always retry to iterate an empty iterator.

Reply:
> of course speedwise it shouldn't matter at all,

I think it may matter if your file starts with 10000 lines of comments, and you ask each one whether it is a byte or string

Reply:
OK, done.